src/manifest_parser.cc

```diff
@@ -311,8 +314,10 @@ bool ManifestParser::ParseEdge(string* err) {
   while (has_indent_token) {
     string key;
     EvalString val;
-    if (!ParseLet(&key, &val, err))
+    if (!ParseLet(&key, &val, err)) {
+      delete env;
```

Comment:
Technically this is correct here, but the reason why is quite subtle.

Also, these new BindingEnv instances are always leaked because there is no single owner for them (several edge->env_ values can point to the same instance).

I suggest you make State be the owner of all BindingEnv instances (for example with a new `std::vector<std::unique_ptr<BindingEnv>> envs_;` member, and just call a method like `state_->NewBindingEnv(env_)` to get a new instance here. In case of parse error, we don't really care about one dangling instance, and we can remove all existing ones in State::Reset() as well.

What do you think about this?

Reply:
Sounds complicated. Wouldn't a `shared_ptr` instead of a pointer be enough?